### ISSUE TITLE
Unable to un-skip tests

### ISSUE THREAD

**Issue by anth0d** [snikch/goodman#32]
I am trying to [add tests for non-2xx codes](http://dredd.readthedocs.io/en/latest/how-to-guides.html#testing-non-2xx-responses-with-swagger) but I am having two problems

First, I am not able to refer to my transaction by name (i.e. `URL > 404 > POST`), so I have given up with that. (I am using Swagger specification, I believe the only example in goodman is using the API Blueprint specification)

So my current strategy is to check for the transaction within the BeforeEach hook, and specify `t.Skip = false` if it is a transaction that should be tested. In fact I am unable to un-skip anything:

Example:

`
	h.BeforeEach(func(t *trans.Transaction) {
		t.Skip = false
                fmt.Println("Don't skip")
        }
`

Output 

> info: Hooks handler stdout: Don't skip
> 
> skip: GET /api/76c75fb2
> info: Hooks handler stderr: 2017/08/08 10:54:35 Shutting down hooks servers

What am I doing wrong?

**Comment by anth0d** [comment]
If I do it in the BeforeAll hook, like so, it works... 

```
	h.BeforeAll(func(t []*trans.Transaction) {
		for _, tr := range t {
			tr.Skip = true
		}
	})
```

**Comment by ddelnano** [comment]
I thought the test suite covers this case but it doesn't look like it does. I will be looking into adding it ensuring that it does in fact fail, then digging into the issue.

**Comment by anth0d** [comment]
I was looking into the issue and believe you'll find that `BeforeAll` and `AfterAll` are working as expected, but the other hooks are not actually modifying the `Transaction`. Probably related to `arg` / `reply` stuff, I am not sure what the hook handlers are doing to be honest, I was going to look into it tomorrow. Let me know how I can help :)

**Comment by ddelnano** [comment]
@antkazam so I am able to prove that this does in fact work properly and I have updated the working example to include a test case of this.  Please see #33 which passes on Travis.

As for the test coverage the go tests should verify that the transactions are in fact modified.  When I mentioned the lack of coverage earlier I was talking about the e2e cucumber tests.

In order to help out what the actual issue is here it would be helpful to see your entire api blueprint file and hook file if you can.  Also you should try and get the example working as I have proven it works as expected.

**Comment by anth0d** [comment]
I'm not seeing it work properly, I believe you can set `t.Skip` but I don't think the updated transaction is getting passed back to Dredd. You can see how it behaves for me - the `fmt.Println` should convince you that the code is at least getting executed.

```
	h.Before("/myapi/v1/pet/{petID} > Get pet by ID > 404 > application/json", func(t *trans.Transaction) {
		fmt.Println("Executing before hook, SHOULD NOT SKIP")
		t.Skip = false
	})
```


```
info: Hooks handler stdout: [GET -> 404] /myapi/v1/pet/91c6ebfbb83a670acf3e15d876c75fb2
info: Hooks handler stdout: Executing before hook, SHOULD NOT SKIP
skip: GET /myapi/v1/pet/91c6ebfbb83a670acf3e15d876c75fb2
```

Notice the "skip" even when `t.Skip = false`.

I have been poking at the code for a bit trying to convince the RPC server to send the updated transactions back to Dredd but have had no luck. I am not sure how it's working for you!

**Comment by ddelnano** [comment]
If this is an open source project please share all code including the API blueprint file.  Have you tried to run the example?  The wiki describes how to run it.

**Comment by anth0d** [comment]
It's a closed source project and we have a Swagger specification, I will try to create a working example and also run the example in the wiki.

As an aside - the problem to me seems a lot like this one, same symptoms from my perspective: https://github.com/snikch/goodman/pull/6

I'm having a hard time understanding why `BeforeAll` is different from `BeforeEach` and `Before`. (i.e. takes references instead of values, calls callback with `args` instead of `reply`)

**Comment by ddelnano** [comment]
What git sha are you building goodman from?  Could it be stale?

**Comment by anth0d** [comment]
`3b65f3e04f92d05e0bdf805fbf137f82f01a1730` - I will let you know when I have more info

**Comment by ddelnano** [comment]
Closing since #35 has been merged and is in the new 0.2.4 release.  Thanks for the help on this!
